site/en/tutorials/interpretability/integrated_gradients.ipynb

```diff
@@ -964,34 +965,42 @@
         "def integrated_gradients(baseline,\n",
         "                         image,\n",
         "                         target_class_idx,\n",
-        "                         m_steps=300,\n",
+        "                         m_steps=50,\n",
         "                         batch_size=32):\n",
-        "  # 1. Generate alphas\n",
-        "  alphas = tf.linspace(start=0.0, stop=1.0, num=m_steps)\n",
+        "  # 1. Generate alphas.\n",
+        "  alphas = tf.linspace(start=0.0, stop=1.0, num=m_steps+1)\n",
         "\n",
-        "  # Accumulate gradients across batches\n",
-        "  integrated_gradients = 0.0\n",
+        "  # Initialize TensorArray outside loop to collect gradients.    \n",
+        "  gradient_batches = tf.TensorArray(tf.float32, size=m_steps+1)\n",
+        "    \n",
+        "  # Iterate alphas range and batch computation for speed, memory efficiency, and scaling to larger m_steps.\n",
+        "  for alpha in tf.range(0, len(alphas), batch_size):\n",
+        "    from_ = alpha\n",
+        "    to = tf.minimum(from_ + batch_size, len(alphas))\n",
+        "    alpha_batch = alphas[from_:to]\n",
         "\n",
-        "  # Batch alpha images\n",
-        "  ds = tf.data.Dataset.from_tensor_slices(alphas).batch(batch_size)\n",
+        "    # 2. Generate interpolated inputs between baseline and input.\n",
+        "    interpolated_path_input_batch = interpolate_images(baseline=baseline,\n",
+        "                                                       image=image,\n",
+        "                                                       alphas=alpha_batch)\n",
         "\n",
-        "  for batch in ds:\n",
+        "    # 3. Compute gradients between model outputs and interpolated inputs.\n",
+        "    gradient_batch = compute_gradients(images=interpolated_path_input_batch,\n",
+        "                                       target_class_idx=target_class_idx)\n",
+        "    \n",
+        "    # Write batch indices and gradients to extend TensorArray.\n",
+        "    gradient_batches = gradient_batches.scatter(tf.range(from_, to), gradient_batch)    \n",
+        "  \n",
+        "  # Stack path gradients together row-wise into single tensor.\n",
+        "  total_gradients = gradient_batches.stack()\n",
         "\n",
-        "    # 2. Generate interpolated images\n",
-        "    batch_interpolated_inputs = interpolate_images(baseline=baseline,\n",
-        "                                                   image=image,\n",
-        "                                                   alphas=batch)\n",
+        "  # 4. Integral approximation through averaging gradients.\n",
+        "  avg_gradients = integral_approximation(gradients=total_gradients)\n",
```

Comment:
Just to be sure I understand:

1. The `TensorArray` is necessary here because you want to get the full list of gradients out of the loop.
1. You want the whole list so that the edges between  batches don't mess with the trapezoidal integration (you loose the trapezoid between the last element of one batch and the first element of the next.
1. It would have been messy to do the `trapz` in the loop, and carry-over the last element to the next batch. 
1. It would have been less precise to do a rectangular integration.

Okay.

Reply:
1. Yes exactly. In addition to working performantly with @tf.function, a side benefit of using TensorArray on this tutorial is highlighting how to add different batch sizes to a TensorArray with scatter() which has been the subject of a few stackoverflow posts. One of the autograph engineers made a really nice analogy between scatter() and a Python lists extend() method and write() to a list's append() method.
2. Yes. Particularly for this version of the tutorial, I felt adding additional code for just this integration method and explanatory text on the endpoint batching issue without covering or visualizing different integration methods like the longer version would be difficult to understand for learners without consulting the longer version of the tutorial and/or alternative resources.
3. bentyeh actually implemented this approach and it is definitely a valid alternative. You can initialize a last_batch tensor, keep track of the last batch by enumerating the batch, have an if statement to check the method, concat the batches to make sure you have the endpoints, do a tf.math.reduce_sum on each gradient batch, add the gradient batches together and divide the total gradients outside the loop by m_steps. This added additional lines of code and explanatory text and the decision to go with an alternative really came down to teachability and readability as opposed to just performance. I felt it was important to highlight to learners that they are averaging across the image gradients of the interpolated path in line with the original paper author's implementation and presentations and keep the focus on IG not the integration methods. In testing, this approach was evaluated with 6 others and the resulting implementation came out significantly faster but this is just a side benefit to teachability and readability considerations and there might even be additional performance optimizations missed as well.
4. Yes the longer version of the tutorial has plots highlighting how Riemann Trapezoidal generally converges quicker and with fewer m_steps as well as how it requires the interval endpoints compared to alternative Riemann methods. This was identified in experiments by the original authors in section 4 of https://arxiv.org/abs/1908.06214 and is how they initially implemented it. However, there are still edge case functions where this might not be the case so it is still important to point out alternatives in the longer version. I have even seen interpretability libraries with Gauss-Legendre quadrature rule integral approximation as well for IG and other gradient-based interpretability methods.